packages/nextjs/src/config/loaders/ast.ts

```diff
@@ -320,3 +328,169 @@ export function removeComments(ast: AST): void {
   const nodesWithComments = ast.find(Node).filter(nodePath => !!nodePath.node.comments);
   nodesWithComments.forEach(nodePath => (nodePath.node.comments = null));
 }
+
+/**
+ * Determines from a given AST of a file, whether the file has a default export or not.
+ */
+export function hasDefaultExport(ast: AST): boolean {
+  const hasDefaultDeclaration = ast.find(ExportDefaultDeclaration).size() > 0;
+  if (hasDefaultDeclaration) {
+    return true;
+  }
+
+  const hasDefaultSpecifier = ast.find(ExportDefaultSpecifier).size() > 0;
+  if (hasDefaultSpecifier) {
+    return true;
+  }
+
+  const hasNamedDefaultExport = ast.find(ExportSpecifier).some(specifier => specifier.node.exported.name === 'default');
+  if (hasNamedDefaultExport) {
+    return true;
+  }
+
+  return false;
+}
+
+/**
+ * Extracts all identifiers from a "RestElement" within a named export declaration statement ("export const val = ...").
+ * "RestElements" are things like "...value" inside a destructuring assignment.
+ *
+ * Example:
+ * "export const { ...restElement1, bar: { ...restElement2 } } = { foo: 1, bar: { baz: 2 } }" --> ["restElement1", "restElement2"]
+ */
+function getExportIdentifiersFromRestElement(restElement: jscsTypes.RestElement): string[] {
+  const identifiers: string[] = [];
+
+  if (Identifier.check(restElement.argument)) {
+    identifiers.push(restElement.argument.name);
+  }
+
+  return identifiers;
+}
+
+/**
+ * Extracts all identifiers from an ArrayPattern within a destructured named export declaration
+ * statement ("export const [val] = [1]").
+ *
+ * This function recursively calls itself and `getExportIdentifiersFromObjectPattern` since destructuring assignments
+ * can be deeply nested with objects and arrays.
+ *
+ * Example:
+ * export const [{ foo: name1 }, [{ bar: [name2]}, name3]] = [{ foo: 1 }, [{ bar: [2] }, 3]];
+ *   --> ["name1", "name2", "name3"]
+ */
+function getExportIdentifiersFromArrayPattern(arrayPattern: jscsTypes.ArrayPattern): string[] {
+  const identifiers: string[] = [];
+
+  arrayPattern.elements.forEach(element => {
+    if (Identifier.check(element)) {
+      identifiers.push(element.name);
+    } else if (ObjectPattern.check(element)) {
+      identifiers.push(...getExportIdentifiersFromObjectPattern(element));
+    } else if (ArrayPattern.check(element)) {
+      identifiers.push(...getExportIdentifiersFromArrayPattern(element));
+    } else if (RestElement.check(element)) {
+      identifiers.push(...getExportIdentifiersFromRestElement(element));
+    }
```

Comment:
On the one hand, my OCD is super pleased with the parallel structure in all three of these conditionals. On the other hand, I think this could be simplified to

```suggestion
    } else if (RestElement.check(element) && Identifier.check(element.argument) {
      identifiers.push(element.argument.name);
    }
```

and then you don't need the function at all (in which case you can obviously ignore my comments about it above).

(Also, is `RestElement.argument` ever anything besides an `Identifier`?)

[Update]: Having read farther, and seeing that this same logic is used three times, I'm rethinking my "Should this even be a function?" question. I think you're right, it should be, assuming you need the `Identifier` check. But I'd make it `getIdentifierFromRestElement` and have it return a string, not an array.

Reply:
Making it return `string | undefined` would also be a bit ugly cause we'd need a check for undefined. Considering things and the cosmetic nature of it, I am going to leave it as is.

Reply:
Can it be `undefined`, though? If we know that it's an `Identifier`, then `name` has to exist. 

Anyway, not a huge deal wither way - I'm fine leaving it.

Reply:
No you're totally right. Had a brainfart. --> 62bba32f44860589700c933a3b48cdbe0bf373ea